src/frontend/Typechecker.ml

```diff
@@ -194,10 +193,12 @@ let stan_math_return_type name arg_tys =
   match name with
   | x when Stan_math_signatures.is_reduce_sum_fn x ->
       Some (UnsizedType.ReturnType UReal)
-  | x when Stan_math_signatures.is_variadic_ode_fn x ->
-      Some (UnsizedType.ReturnType (UArray UVector))
-  | x when Stan_math_signatures.is_variadic_dae_fn x ->
-      Some (UnsizedType.ReturnType (UArray UVector))
+  | x when Stan_math_signatures.is_stan_math_variadic_function_name x ->
+      Some
+        (UnsizedType.ReturnType
+           (Hashtbl.find_exn Stan_math_signatures.stan_math_variadic_signatures
+              x )
+             .return_type )
   | _ ->
       SignatureMismatch.matching_stanlib_function name arg_tys
       |> match_to_rt_option
```

Comment:
How about
```ocaml
let stan_math_return_type name arg_tys =
  match (Hashtbl.find Stan_math_signatures.stan_math_variadic_signatures name) with
  | _ when Stan_math_signatures.is_reduce_sum_fn name ->
      Some (UnsizedType.ReturnType UReal)
  | Some { return_type ; _ } -> Some(UnsizedType.ReturnType return_type)
  | None ->
      SignatureMismatch.matching_stanlib_function name arg_tys
      |> match_to_rt_option
```
and maybe move the whole function to SignatureMismatch since it's not needed here?

Reply:
Agreed on the refactor, I prefer to keep it in Typechecker (in part because I am still hopeful for something like #1182, where it makes the most sense to encapsulate it inside of Typechecking)